src/Engine/Army.java

```diff
@@ -265,50 +292,43 @@ protected boolean awbwDeniesCharge()
   @Override
   public GameEventQueue receiveBattleEvent(final BattleSummary summary)
   {
-    if( awbwDeniesCharge() )
-      return null;
+    // The lambdas demand the args be effectively final, so we're duplicating checks today
+    boolean amAttacking = this == summary.attacker.CO.army;
+    if( !amAttacking && this != summary.defender.CO.army )
+      return null; // not attacking and not the defender -> not involved
 
     // We only care who the units belong to, not who picked the fight.
-    UnitDelta minion = null;
-    UnitDelta enemy = null;
-    boolean isCounter = false;
-    if( this == summary.attacker.CO.army )
+    final UnitDelta minion;
+    final UnitDelta enemy;
+    final boolean isCounter = !amAttacking;
+    if( amAttacking )
     {
       minion = summary.attacker;
       enemy = summary.defender;
-      isCounter = false;
     }
-    if( this == summary.defender.CO.army )
+    else
     {
       minion = summary.defender;
       enemy = summary.attacker;
-      isCounter = true;
     }
 
-    if( minion != null && enemy != null )
+    Function<Commander, Integer> chargeEvaluator = (co) -> co.calculateCombatCharge(minion, enemy, isCounter);
+    switch (gameRules.tagMode)
     {
-      switch (gameRules.tagMode)
-      {
-        case AWBW:
-        {
-          final int primaryCharge = cos[0].calculateCombatCharge(minion, enemy, isCounter);
-          cos[0].modifyAbilityPower(primaryCharge);
-          for( int i = 1; i < cos.length; ++i )
-          {
-            final int tagCharge = cos[i].calculateCombatCharge(minion, enemy, isCounter);
-            final int tagDivisor = 2;
-            cos[i].modifyAbilityPower(tagCharge / tagDivisor);
-          }
-        }
-          break;
-        case Team_Merge:
-        case OFF:
+      case AWBW:
+        awbwTagsCharge(chargeEvaluator);
+        break;
+      case Persistent:
+        persistentTagsCharge(chargeEvaluator);
+        break;
+      case Team_Merge:
+      case OFF:
+        if( minion.CO.canAcceptCharge() )
```

Comment:
Should we just do the `canAcceptCharge()` check inside `modifyAbilityPower()` to avoid needing to remember to check it elsewhere?

Reply:
I'm gonna say "no", since I think non-standard providers of charge should still work while a power is active.
And in any case, it's not called `maybeModifyAbilityPower()`

Reply:
Ugh. Gross.